File: ios/Sources/GameViewModel.swift
//  GameViewModel.swift
//  라비린스 — 화면 상태 관리. 서버 스냅샷을 받아 SwiftUI 뷰에 게시한다.
//
//  내 좌석(mySeat)은 서버 스냅샷의 players[*].currentTarget 가 채워진 좌석으로 추론한다
//  (서버가 본인 시점에서만 목표 보물을 노출하기 때문). 게스트/봇뷰어 대비.

import Foundation
import Combine

@MainActor
final class GameViewModel: ObservableObject {
    enum Screen { case lobby, room, game, result }

    @Published var screen: Screen = .lobby
    @Published var roomInfo: RoomInfo?
    @Published var snapshot: LabyrinthSnapshot?
    @Published var turn: TurnInfo?
    @Published var gameOver: GameOver?
    @Published var errorMessage: String?
    @Published var connected = false

    // 삽입 단계 임시 선택(회전 미리보기)
    @Published var pendingRotation: Int = 0
    @Published var selectedInsertion: Int?

    private let service = SocketService.shared

    /// 내 좌석 추론: currentTarget 가 non-nil 인 플레이어 = 나.
    var mySeat: Int? {
        snapshot?.players.first(where: { $0.currentTarget != nil })?.index
    }
    var isMyTurn: Bool {
        guard let s = snapshot, let me = mySeat else { return false }
        return s.currentPlayer == me && s.phase != .finished
    }
    var spareRotationPreview: Int {
        guard let spare = snapshot?.spare else { return 0 }
        return selectedInsertion != nil ? pendingRotation : spare.rotation
    }

    func bind() {
        service.onConnected = { [weak self] in self?.connected = true }
        service.onRoom = { [weak self] in self?.roomInfo = $0; self?.screen = .room }
        service.onState = { [weak self] snap in
            self?.snapshot = snap
            if snap.phase == .finished { return }
            if self?.screen != .game { self?.screen = .game }
        }
        service.onTurn = { [weak self] in
            self?.turn = $0
            // 새 턴이면 회전 선택 초기화
            self?.selectedInsertion = nil
            self?.pendingRotation = self?.snapshot?.spare.rotation ?? 0
        }
        service.onGameOver = { [weak self] in self?.gameOver = $0; self?.screen = .result }
        service.onError = { [weak self] in self?.errorMessage = $0 }
        service.onCreated = { _ in }
        service.onJoined = { _ in }
    }

    // MARK: 액션

    func connect(token: String?, nickname: String?) {
        bind()
        service.connect(token: token, nickname: nickname)
    }
    func createRoom(maxPlayers: Int) { service.createRoom(maxPlayers: maxPlayers) }
    func joinRoom(code: String) { service.joinRoom(code: code.uppercased()) }
    func addBot() { service.addBot() }
    func start() { service.start() }
    func quickMatch() { service.quickMatch() }
    func leave() { service.leave(); screen = .lobby; snapshot = nil; roomInfo = nil }

    /// 삽입 지점 탭: 처음 탭이면 선택, 같은 지점 재탭이면 회전.
    func tapInsertion(_ id: Int) {
        guard isMyTurn, snapshot?.phase == .insert else { return }
        if selectedInsertion == id {
            pendingRotation = (pendingRotation + 1) % 4
        } else {
            selectedInsertion = id
            pendingRotation = snapshot?.spare.rotation ?? 0
        }
    }
    /// 선택한 삽입 확정
    func confirmInsertion() {
        guard isMyTurn, let id = selectedInsertion else { return }
        service.insert(insertionId: id, rotation: pendingRotation)
        selectedInsertion = nil
    }
    func rotateSpare() {
        guard isMyTurn, snapshot?.phase == .insert else { return }
        pendingRotation = (pendingRotation + 1) % 4
    }
    func moveTo(_ index: Int) {
        guard isMyTurn, snapshot?.phase == .move else { return }
        service.move(to: index)
    }
}

File: android/app/src/main/java/com/jiny/labyrinthonline/GameViewModel.kt
package com.jiny.labyrinthonline

import android.os.Build
import android.os.Handler
import android.os.Looper
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel

/*
 * 화면 상태 관리. 서버 스냅샷을 받아 Compose state 로 노출한다.
 * 내 좌석(mySeat)은 currentTarget 가 채워진 좌석으로 추론(서버가 본인 시점에만 노출).
 */
enum class Screen { LOBBY, ROOM, GAME, RESULT }

class GameViewModel : ViewModel() {
    private val service = SocketService()
    private val main = Handler(Looper.getMainLooper())

    var screen by mutableStateOf(Screen.LOBBY); private set
    var roomInfo by mutableStateOf<RoomInfo?>(null); private set
    var snapshot by mutableStateOf<LabyrinthSnapshot?>(null); private set
    var turn by mutableStateOf<TurnInfo?>(null); private set
    var gameOver by mutableStateOf<GameOver?>(null); private set
    var errorMessage by mutableStateOf<String?>(null)
    var connected by mutableStateOf(false); private set
    val chatMessages = mutableStateListOf<ChatMessage>()
    var queuing by mutableStateOf(false); private set
    var queueWaiting by mutableStateOf(0); private set
    var queueNeed by mutableStateOf(0); private set
    var paused by mutableStateOf(false); private set

    // 삽입 단계 임시 선택
    var selectedInsertion by mutableStateOf<Int?>(null); private set
    var pendingRotation by mutableStateOf(0); private set

    val mySeat: Int?
        get() = snapshot?.players?.firstOrNull { it.currentTarget != null }?.index

    val isMyTurn: Boolean
        get() {
            val s = snapshot ?: return false
            val me = mySeat ?: return false
            return s.currentPlayer == me && s.phase != Phase.FINISHED
        }

    fun connect(token: String?, nickname: String?) {
        service.onConnected = { main.post { connected = true } }
        service.onRoom = { r -> main.post { roomInfo = r; screen = Screen.ROOM } }
        service.onState = { snap ->
            main.post {
                snapshot = snap
                queuing = false
                if (snap.phase != Phase.FINISHED && screen != Screen.GAME) screen = Screen.GAME
            }
        }
        service.onQueued = { _, waiting, need -> main.post { queuing = true; queueWaiting = waiting; queueNeed = need } }
        service.onPaused = { main.post { paused = true } }
        service.onResumed = { main.post { paused = false } }
        service.onAborted = { main.post { errorMessage = "GAME_ABORTED"; leave() } }
        service.onTurn = { t ->
            main.post {
                turn = t
                selectedInsertion = null
                pendingRotation = snapshot?.spare?.rotation ?: 0
            }
        }
        service.onGameOver = { g -> main.post { gameOver = g; screen = Screen.RESULT } }
        service.onChatMessage = { m -> main.post {
            chatMessages.add(m)
            if (chatMessages.size > 100) chatMessages.removeAt(0)
        } }
        service.onError = { code -> main.post { errorMessage = code } }
        service.connect(token, nickname ?: ("Android-" + Build.MODEL))
    }

    // 액션
    fun createRoom(maxPlayers: Int) { service.createRoom(maxPlayers) }
    fun joinRoom(code: String) { service.joinRoom(code.uppercase()) }
    fun addBot() { service.addBot() }
    fun start() { service.start() }
    fun quickMatch(size: Int) { queuing = true; service.quickMatch(size) }
    fun cancelQuick() { queuing = false; service.cancelQuick() }
    fun leave() { service.leave(); screen = Screen.LOBBY; snapshot = null; roomInfo = null; queuing = false; paused = false }

    fun tapInsertion(id: Int) {
        if (!isMyTurn || snapshot?.phase != Phase.INSERT) return
        if (selectedInsertion == id) {
            pendingRotation = (pendingRotation + 1) % 4
        } else {
            selectedInsertion = id
            pendingRotation = snapshot?.spare?.rotation ?: 0
        }
    }
    fun rotateSpare() {
        if (!isMyTurn || snapshot?.phase != Phase.INSERT) return
        pendingRotation = (pendingRotation + 1) % 4
    }
    fun confirmInsertion() {
        val id = selectedInsertion ?: return
        if (!isMyTurn) return
        service.insert(id, pendingRotation)
        selectedInsertion = null
    }
    fun moveTo(index: Int) {
        if (!isMyTurn || snapshot?.phase != Phase.MOVE) return
        service.move(index)
    }
    fun sendChat(text: String) {
        val t = text.trim()
        if (t.isNotEmpty()) service.sendChat(t)
    }

    override fun onCleared() { service.disconnect() }
}
